Calendar Block: Add color supports and polish styles

t-hamano opened the pull request:
Fix: #40989

## What?
This PR adds color block support to the calendar block and updates the style to match a variety of themes.

## Why?
This block has hard-coded text color and border color.
This causes problems that the text is difficult to read and the design doesn't match the other blocks, depending on the background color of the content.

## How?
I've made the following changes based on a table block (using the same `table` tag):

- Remove hard-coded colors
- Border color inherit text color
- Make `thead` bottom border thicker
- Add text and link color with block support

## Testing Instructions

- Check out this branch, and build sources.
- Insert a calendar block in the site editor ( or the post editor ).
- Change the global style and see how the appearance of the calendar block changes.
- If possible, check with classic themes or other block themes as well.

## Screenshots or screencast <!-- if applicable -->

https://user-images.githubusercontent.com/54422211/176364901-def33894-5464-4493-ace7-2565912b27ff.mp4

t-hamano committed:
Calender Block: Add color supports and polish styles

t-hamano commented on packages/block-library/src/calendar/block.json:
```diff
@@ -16,7 +16,14 @@
+		"align": true,
+		"color": {
+			"link": true,
+			"background": false,
```
The table block supports background color, should it also support it in the calendar block?
In my opinion, I think support is unnecessary.

carolinan replied:
I think it should be enabled because it would give everyone more design options.

scruffian replied:
I agree

ramonjd replied:
Sorry, late to the party. I had been toying around with adding color support for the calendar block over in https://github.com/WordPress/gutenberg/pull/43299

I'll close my PR in favour of this one. 

Just wanted to mention something I discovered in relation to background color support:

When selecting a preset color, followed by a custom color, the block doesn't rerender.

I've pinpointed it to the use of [useDebounce in the ServerSideRender component](916abc45e6682f4bdb70888aa41e98aa395/packages/server-side-render/src/server-side-render.js#L145).

The debounced function is called, but it doesn't fire. I suspect that it's cancelled too early [here](https://github.com/WordPress/gutenberg/blob/33d84b036592a5bf31af05b7710f3b2b14163dc4/packages/compose/src/hooks/use-debounce/index.js). 

It's a strange equality check bug that I haven't worked out yet.

I tested on this branch:


https://user-images.githubusercontent.com/6458278/185529470-2f59fe69-5422-46b4-9a74-cdbe135da014.mp4




It occurs every so often, but not all the time so might be acceptable.

But if I enable`"gradients": true` in color supports, the effect is more pronounced:


https://user-images.githubusercontent.com/6458278/185529396-9c5dfe29-f142-4d63-badb-fa8bde79ee6e.mp4


Lesson: don't enable "gradient" for now 😄

carolinan commented:
@WordPress/block-themers Any feedback on this calendar block PR?
I personally prefer the style after the PR is applied, but will the changes, especially the removal of the background color, be too much for existing themes?

carolinan commented:
Pinging @[markhowellsmead](https://github.com/markhowellsmead) as well in case you want to review this CSS change.

scruffian commented:
> will the changes, especially the removal of the background color, be too much for existing themes?

I am concerned about the loss of the background color too. Maybe we can scope this back to changes which won't have a significant impact on existing blocks?

I also think the calendar caption should have a `wp-element-caption` class added to it, so it can be styled consistent with other captions. This is another example where the HTML Walker class will be very useful: https://github.com/WordPress/gutenberg/pull/42485

t-hamano commented:
I tried to support background but hit a wall.

First, if we simply opt for background with block support, the style will be applied to the block itself and will look like this:

![calendar_1](https://user-images.githubusercontent.com/54422211/183633960-bebd8cc3-a9b0-4a24-bb1d-4fedadd8c3c4.png)

But most people would expect the background color to be applied to the table tag inside, just as the table block is:
![calendar_2](https://user-images.githubusercontent.com/54422211/183633989-e02a466a-26b2-4a2e-9d6d-aefed8eecdc1.png)

In this case, `__ExperimentalSkipSerialization` must be enabled and the style must be explicitly generated in the `render_block` callback function.
There is no problem If only text and background colors are supported, but there is a problem with link color support.

If a link color is applied in a static block, `render_block` / `pre_render_block` filter will execute [wp_render_elements_support](https://github.com/WordPress/gutenberg/blob/49e57d4bf0f3c6574e4c3369cea9acd04d1aea4f/lib/block-supports/elements.php#L25) and [wp_render_elements_support_styles](https://github.com/WordPress/gutenberg/blob/49e57d4bf0f3c6574e4c3369cea9acd04d1aea4f/lib/block-supports/elements.php#L89). Then, a class name such as `wp-elements-XXX` is given and the following inline style will be output:

![link](https://user-images.githubusercontent.com/54422211/183635488-c19ff9e3-18df-4cc5-abe5-40f469f8969a.png)

However, this process is not executed on dynamic blocks that support link colors and have skipped serialization. Therefore, we need to generate the class name and inline styles within the `render_callback` function of the calendar block. Or, prepare a similar function for dynamic blocks.

Does anyone have a better way to do this? 🙏

pagelab commented:
Please also remove the hard-coded `text-decoration` property in the CSS files for this block. It's just unnecessary.

draganescu commented:
@t-hamano the way you described your approaches sounds pretty good to me!

ramonjd commented:
> I am concerned about the loss of the background color too. Maybe we can scope this back to changes which won't have a significant impact on existing blocks?

In https://github.com/WordPress/gutenberg/pull/43299 I rather inelegantly got around this using `&:not(.has-text-color)` and `&:not(.has-background)`

https://github.com/WordPress/gutenberg/pull/43299/files#diff-ed2baf015a5fd6164b7e4aa9596ae4f8ca30d10d16720cb43625b138487ee5ceR31

There's probably a better way than my hack though.

t-hamano committed:
Add background-color support, add color and style dynamic rendering, remove text-decoration

t-hamano commented:
I have added the following brush ups: 

### Backward compatibility for the header background color

If both text and background color are not set, the previous gray header background color is retained.

https://github.com/WordPress/gutenberg/blob/aaa96dc2e3c215f991e166c61dd43e6933f35dba/packages/block-library/src/calendar/style.scss#L10-L18

### Background color support

I think text and background colors should only apply to tables, just as table blocks do.
Therefore, I opted for `__experimentalSkipSerialization` and rendered the required styles and classes in the render callback function.

The code is based on [the search block](https://github.com/WordPress/gutenberg/blob/ca14481dabfd2f409baa69433f5f8e524f46fbf1/packages/block-library/src/search/index.php#L38-L39), but if [the HTML Walker class](https://github.com/WordPress/gutenberg/pull/42485) is adopted, it will be simpler to write.

### Remove text-decoration style

`text-decoration: underline` is hard coded in the post link.
Although this point is mentioned in [this comment](https://github.com/WordPress/gutenberg/pull/42029#issuecomment-1215776196), I think the theme should be responsible for this and removed it.

### About skip serialization

> In this case, __ExperimentalSkipSerialization must be enabled and the style must be explicitly generated in the render_block callback function.
> There is no problem If only text and background colors are supported, but there is a problem with link color support.
> 
> If a link color is applied in a static block, render_block / pre_render_block filter will execute [wp_render_elements_support](https://github.com/WordPress/gutenberg/blob/49e57d4bf0f3c6574e4c3369cea9acd04d1aea4f/lib/block-supports/elements.php#L25) and [wp_render_elements_support_styles](https://github.com/WordPress/gutenberg/blob/49e57d4bf0f3c6574e4c3369cea9acd04d1aea4f/lib/block-supports/elements.php#L89). Then, a class name such as wp-elements-XXX is given and the following inline style will be output:

I had assumed that `__ExperimentalSkipSerialization` could only specify boolean values about the above problem.

However, I learned that `__ExperimentalSkipSerialization` can specify properties to be skipped, so I excluded the link property.
I believe that all color support has worked as expected.

### Screencast

https://user-images.githubusercontent.com/54422211/185797183-6b20f67b-d66e-4ff6-9957-8437c1dbf0b2.mp4

t-hamano committed:
Fix php lint error

t-hamano commented:
> I also think the calendar caption should have a wp-element-caption class added to it, so it can be styled consistent with other captions. 

Regarding the above proposal, I would like to push an additional issue and PR if this PR is merged.

scruffian commented:
Thanks for your work here. This is looking almost ready, just one question about the border you added.

t-hamano commented:
@scruffian 
Thank you for checking!

> just one question about the border you added.

Can you please elaborate on this mention if there is anything I should add to address?

scruffian commented on packages/block-library/src/calendar/style.scss:
```diff
@@ -2,31 +2,22 @@
-	table caption {
-		color: #40464d;
+	thead {
+		border-bottom: 3px solid;
```
Sorry it looks like my comment didn't send the first time:

What's the reason for adding this? I think we should be leaning towards removing styles rather than adding them

t-hamano replied:
I would like you to watch this video regarding the style comparison with table blocks in my previous comment, The table block has a similar border under `thead`.
I have made the same changes to the calendar block for style consistency, but do you think this is unnecessary?

scruffian replied:
My concern is that this will change the appearance of the calendar blocks that people are already using, so I'd rather err on the side of caution and not add more styles.

aaronrobertshaw replied:
+1 to not introducing any new styles.

t-hamano replied:
Ok, I'm going to remove this style 👍

t-hamano replied:
Fixed by 8b4e335

aaronrobertshaw commented on packages/block-library/src/calendar/block.json:
```diff
@@ -16,7 +16,15 @@
+		"align": true,
+		"color": {
+			"link": true,
+			"__experimentalSkipSerialization": [ "text", "background" ],
```
Maybe we could leverage `__experimentalSelector` within this support's config to help theme.json and global styles to target the correct elements?

aaronrobertshaw commented on packages/block-library/src/calendar/index.php:
```diff
@@ -40,11 +40,17 @@ function render_block_core_calendar( $attributes ) {
 		}
 	}
 
+	$inline_styles = styles_for_block_core_calendar( $attributes );
+	$color_classes = get_color_classes_for_block_core_calendar( $attributes );
```
We should probably leverage the style engine to generate all the color class names and inline styles. See [`lib/block-supports/colors.php`](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/colors.php) for how the block supports do this when serialization isn't skipped.

aaronrobertshaw commented on packages/block-library/src/calendar/index.php:
```diff
@@ -69,6 +75,66 @@ function register_block_core_calendar() {
+		$table_styles[] = sprintf( 'background-color: %s;', esc_attr( $attributes['style']['color']['background'] ) );
+	}
+
+	return ! empty( $table_styles ) ? sprintf( 'style="%s"', safecss_filter_attr( implode( ' ', $table_styles ) ) ) : '';
```
If we continue to manually generate the styles here free of the style engine, perhaps we can move the `esc_attr` calls here to wrap `safecss_filter_attr` to help ensure any future styles added to this function don't miss proper escaping.

aaronrobertshaw commented on packages/block-library/src/calendar/index.php:
```diff
@@ -69,6 +75,66 @@ function register_block_core_calendar() {
+ *
+ * @return string The color classnames to be applied to the block elements.
+ */
+function get_color_classes_for_block_core_calendar( $attributes ) {
```
We might be able to reduce some of the duplicate logic in class generation if we leverage the style engine.

aaronrobertshaw commented on packages/block-library/src/calendar/style.scss:
```diff
@@ -2,31 +2,22 @@
-	a {
-		text-decoration: underline;
+		// Keep the hard-coded header background color for backward compatibility.
+		&:not(.has-text-color):not(.has-background) th {
```
I don't think increasing the specifity here is a great option to apply a default. Maybe we could leverage `:where` instead?

t-hamano replied:
Fixed by 5adb2a6

aaronrobertshaw reviewed (changes requested):
Thanks for working on calendar colors @t-hamano 🙇 

Unfortunately, this PR isn't working very well for me. 

With the current approach, we're skipping serialization to apply styles to inner elements, but nothing has been changed for theme.json and Global Styles to target those same elements. Adding a feature-level selector to the color supports config via `__experimentalSelector` might be of assistance here.

In addition to this, the colors are inconsistently applied across the block editor, site editor, and front end.

If we can fix the application of color styles for theme.json and global styles, address some specificity concerns, and leverage the style engine, I'm hopeful this PR isn't that far off being ready.

Here's a video of what I'm seeing:


https://user-images.githubusercontent.com/60436221/189293366-4cafcc0e-efaa-4b23-9dc7-68a4f5ed067a.mp4

t-hamano committed:
Merge branch 'trunk' into try/block-calendar

t-hamano committed:
Remove unnecessary border styles

t-hamano committed:
Lower specificity of background color for backward compatibility

t-hamano committed:
Enable color support by global style

t-hamano committed:
Use style-engine for inline CSS

t-hamano committed:
USe style-engine for class names

t-hamano commented:
Thank you for your kind advice, @aaronrobertshaw !
I have made the following improvements:

- Remove `border-bottom` style from `thead`
- Reduced `thead` background color specificity for backward compatibility
- Use style-engine for class names and inline CSS (with `__experimentalSelector`)
- Enable color support by global style (with `(gutenberg|wp)_style_engine_get_styles`)

https://user-images.githubusercontent.com/54422211/189478072-70ce44aa-d7b5-41d1-bf5c-476b632e0c09.mp4

Generally works as expected, but there is a problem with the gray header background color for backward compatibility.
When a color is applied via global style, `has-text-color` and `has-background` classes aren't applied to the table, so the gray background color is retained.
I think the only way to solve this is to remove this backward compatibility, what do you think?

aaronrobertshaw commented on packages/block-library/src/calendar/style.scss:
```diff
@@ -2,31 +2,18 @@
 		border-collapse: collapse;
-	}
-
-	table th {
-		font-weight: 400;
-		background: $gray-300;
-	}
```
Should we also be maintaining the font-weight for backwards compatibility? Same for the `tbody td` border color, `table tbody, table caption` text color, and table link text decoration?

aaronrobertshaw reviewed (commented):
Thanks for the continued effort @t-hamano 🙇 

I think this PR is really starting to take shape!

After reviewing the latest changes, it appears as though we are currently split only half using the style engine, i.e. for inline styles only. I think being consistent and using it for all the block support generated styles would be beneficial and automatically pick up color support changes in the future.

I've added a diff below that updates the use of the style engine, among other fixes, and I think this cleans things up a little.

> Generally works as expected, but there is a problem with the gray header background color for backward compatibility.
When a color is applied via global style, has-text-color and has-background classes aren't applied to the table, so the gray background color is retained.
I think the only way to solve this is to remove this backward compatibility, what do you think?

Luckily, I don't think that is the only way to solve this. We could target the `th` with the feature-level selectors as well so that they'll trump the `:where` rule you've already added. Then with a few extra tweaks to the styles, we can smooth out the consequences for individual blocks. I also noticed a bug in that setting a text color prevented the default background from applying.

I've included the suggestions above in the final diff below. The demo video below is the result I get when applying the diff.

<details>
<summary><strong>Diff illustrating suggestions</strong></summary>

```diff
diff --git a/packages/block-library/src/calendar/block.json b/packages/block-library/src/calendar/block.json
index 747bfd67c8..d1f2e0b604 100644
--- a/packages/block-library/src/calendar/block.json
+++ b/packages/block-library/src/calendar/block.json
@@ -24,7 +24,7 @@
 				"background": true,
 				"text": true
 			},
-			"__experimentalSelector": "table"
+			"__experimentalSelector": "table, table th"
 		},
 		"typography": {
 			"fontSize": true,
diff --git a/packages/block-library/src/calendar/editor.scss b/packages/block-library/src/calendar/editor.scss
new file mode 100644
index 0000000000..b12256cb74
--- /dev/null
+++ b/packages/block-library/src/calendar/editor.scss
@@ -0,0 +1,13 @@
+.editor-styles-wrapper {
+	.wp-block-calendar {
+		table {
+			&.has-background th {
+				background-color: inherit;
+			}
+
+			&.has-text-color th {
+				color: inherit;
+			}
+		}
+	}
+}
diff --git a/packages/block-library/src/calendar/index.php b/packages/block-library/src/calendar/index.php
index eac45b486b..c3b89f67e8 100644
--- a/packages/block-library/src/calendar/index.php
+++ b/packages/block-library/src/calendar/index.php
@@ -40,29 +40,25 @@ function render_block_core_calendar( $attributes ) {
 		}
 	}
 
-	$color_classes = array();
+	$color_block_styles = array();
 
-	if ( ! empty( $attributes['textColor'] ) ) {
-		$color_classes[] = sprintf( 'has-text-color has-%s-color', $attributes['textColor'] );
-	}
-	if ( ! empty( $attributes['backgroundColor'] ) ) {
-		$color_classes[] = sprintf( 'has-background has-%s-background-color', $attributes['backgroundColor'] );
-	}
+	// Text color.
+	$preset_text_color          = array_key_exists( 'textColor', $attributes ) ? "var:preset|color|{$attributes['textColor']}" : null;
+	$custom_text_color          = _wp_array_get( $attributes, array( 'style', 'color', 'text' ), null );
+	$color_block_styles['text'] = $preset_text_color ? $preset_text_color : $custom_text_color;
 
-	$inline_styles = '';
+	// Background Color.
+	$preset_background_color          = array_key_exists( 'backgroundColor', $attributes ) ? "var:preset|color|{$attributes['backgroundColor']}" : null;
+	$custom_background_color          = _wp_array_get( $attributes, array( 'style', 'color', 'background' ), null );
+	$color_block_styles['background'] = $preset_background_color ? $preset_background_color : $custom_background_color;
 
-	if ( ! empty( $attributes['style'] ) ) {
-		$styles = gutenberg_style_engine_get_styles( $attributes['style'] );
-		if ( ! empty( $styles['css'] ) ) {
-			$inline_styles = " style=\"${styles['css']}\"";
-		}
-		if ( ! empty( $styles['classnames'] ) ) {
-			$color_classes[] = $styles['classnames'];
-		}
-	}
+	// Generate color styles and classes.
+	$styles        = gutenberg_style_engine_get_styles( array( 'color' => $color_block_styles ), array( 'convert_vars_to_classnames' => true ) );
+	$inline_styles = empty( $styles['css'] ) ? '' : sprintf( ' style="%s"', esc_attr( $styles['css'] ) );
 
-	$calendar = str_replace( '<table', '<table' . $inline_styles, get_calendar( true, false ) );
-	$calendar = str_replace( 'class="wp-calendar-table', 'class="wp-calendar-table ' . implode( ' ', $color_classes ), $calendar );
+	// Apply color classes and styles to the calendar.
+	$calendar      = str_replace( '<table', '<table' . $inline_styles, get_calendar( true, false ) );
+	$calendar      = str_replace( 'class="wp-calendar-table', 'class="wp-calendar-table ' . esc_attr( $styles['classnames'] ), $calendar );
 
 	$wrapper_attributes = get_block_wrapper_attributes();
 	$output             = sprintf(
diff --git a/packages/block-library/src/calendar/style.scss b/packages/block-library/src/calendar/style.scss
index 3095e3f50a..f28900c0ae 100644
--- a/packages/block-library/src/calendar/style.scss
+++ b/packages/block-library/src/calendar/style.scss
@@ -12,8 +12,16 @@
 		border-collapse: collapse;
 
 		// Keep the hard-coded header background color for backward compatibility.
-		&:where(:not(.has-text-color):not(.has-background)) th {
+		&:where(:not(.has-background)) th {
 			background: $gray-300;
 		}
+
+		&.has-background th {
+			background-color: inherit;
+		}
+
+		&.has-text-color th {
+			color: inherit;
+		}
 	}
 }
diff --git a/packages/block-library/src/editor.scss b/packages/block-library/src/editor.scss
index edb66416af..3af72b2448 100644
--- a/packages/block-library/src/editor.scss
+++ b/packages/block-library/src/editor.scss
@@ -4,6 +4,7 @@
 @import "./block/editor.scss";
 @import "./button/editor.scss";
 @import "./buttons/editor.scss";
+@import "./calendar/editor.scss";
 @import "./categories/editor.scss";
 @import "./code/editor.scss";
 @import "./columns/editor.scss";

```

</details>

<details>
<summary><strong>Demo video</strong></summary>

<video src="https://user-images.githubusercontent.com/60436221/189623633-cf889c44-f54d-42c0-91fe-b5905e1c6517.mp4"/>
</details>

---

Lastly, is there anything else we can do to maintain backwards compatibility for the other default styles removed in this PR? I've run out of time for today to digger any deeper sorry.

Hope this helps; good luck!

t-hamano committed:
Style engine covers more, improv backward compatibility

t-hamano committed:
Add more backward compatibility

t-hamano commented:
I appreciate your detailed advice, @aaronrobertshaw!

Again, I would like to summarize the specifications of the calendar block in `trunk` and the behavior expected by this PR.
(If there is any misunderstanding, please point it out.)

## The calendar block in the trunk

![calendar](https://user-images.githubusercontent.com/54422211/189692162-efd9e61f-be51-4bbd-8b5e-d2308f816aef.png)

- body (`th`, `td`) and caption have `#40464d` text color
- `th` and `td` have `#dddddd` border-color
- `th` has `#dddddd` background color
- `th` has normal font-weight
- Link has `test-decoration: underline;` (As mentioned by [this comment](https://github.com/WordPress/gutenberg/pull/42029#issuecomment-1215776196), it might not be originally necessary.)

## The behavior expected by this PR

- Keep the previous style as long as the color is not changed.
- Changing the background color affects in the entire table, including the header.
- Changing the text color affects all text except links and the border color of all cells.
- Individual block color settings override global styles.
- Remove hard-coded `text-decoration: underline;` on links

t-hamano committed:
Fix php lint error

t-hamano committed:
Merge branch 'trunk' into try/block-calendar

t-hamano committed:
Fix: global style text color isn't applied to td

t-hamano committed:
Fix: global style text color isn't applied to caption

aaronrobertshaw reviewed (commented):
I appreciate your patience with this PR @t-hamano. 🙇 

Unfortunately, the latest changes adding more backwards compatibility, removing some of the previously suggested CSS rules, and including `th` and `caption` in the feature level selector have resulted in a few regressions.

During my latest round of testing, the link color still functions well, but there were issues with the background and text colors, as well as the caption.

- Global Styles set background colors don't override the default `th` background on the frontend
- Individual block text and background colors now don't override the Global Styles set values
- The background color doesn't get applied to the caption when setting it only for individual blocks

After some more fiddling with our combination of feature level selectors and styles I believe I have it working and ticking all the boxes you outlined in your [last comment](https://github.com/WordPress/gutenberg/pull/42029#issuecomment-1243902622). The suggested diff is included below.

<details>
<summary><strong>Click for suggested changes</strong></summary>

```diff
diff --git a/packages/block-library/src/calendar/block.json b/packages/block-library/src/calendar/block.json
index 89afb26410..2accd7142c 100644
--- a/packages/block-library/src/calendar/block.json
+++ b/packages/block-library/src/calendar/block.json
@@ -24,7 +24,7 @@
 				"background": true,
 				"text": true
 			},
-			"__experimentalSelector": "table, th, td, caption"
+			"__experimentalSelector": "table, th"
 		},
 		"typography": {
 			"fontSize": true,
diff --git a/packages/block-library/src/calendar/style.scss b/packages/block-library/src/calendar/style.scss
index ddeb10d421..969d1aafec 100644
--- a/packages/block-library/src/calendar/style.scss
+++ b/packages/block-library/src/calendar/style.scss
@@ -11,21 +11,16 @@
 		font-weight: 400;
 	}
 
+	caption {
+		background-color: inherit;
+	}
+
 	table {
 		width: 100%;
 		border-collapse: collapse;
 
-		// Keep the hard-coded header background color for backward compatibility.
-		&:where(:not(.has-background)) th {
-			background: $gray-300;
-		}
-
 		&:where(:not(.has-text-color)) {
-			// Keep the hard-coded body and caption color for backward compatibility.
-			tbody,
-			caption {
-				color: #40464d;
-			}
+			color: #40464d;
 
 			// Keep the hard-coded border color for backward compatibility.
 			th,
@@ -33,5 +28,18 @@
 				border-color: $gray-300;
 			}
 		}
+
+		&.has-background th {
+			background-color: inherit;
+		}
+
+		&.has-text-color th {
+			color: inherit;
+		}
 	}
 }
+
+// Keep the hard-coded header background color for backward compatibility.
+:where(.wp-block-calendar table:not(.has-background) th) {
+	background: $gray-300;
+}

```
</details>

- I removed the recently added `td` and `caption` from the feature level selectors
- Added a style for the `caption` so it would inherit any background applied to the table
- Simplified the default color fallback. Although it lowers the specificity, I believe this was the intent by introducing the `:where` selector
- Restored the previously suggested CSS rules to inherit background color and text color when set at the individual block level
- Moved the default `th` background color so it was contained within a single `:where` selector allowing global styles to override it

The styles might still be able to be cleaned up further but I've run out of time tonight.

Below are demo videos from before my latest suggested changes and after.

| Before (Current PR) | After (With changes Above) |
|---|---|
| <video src="https://user-images.githubusercontent.com/60436221/190367602-35f3cb98-8ee1-418d-97be-63d8dd8d3eb1.mp4"/> | <video src="https://user-images.githubusercontent.com/60436221/190367659-37a43d92-76cd-4a47-aec3-31b6866724e9.mp4"/> |

t-hamano commented:
Thanks for the further suggestions, @aaronrobertshaw !

I was just wrestling with the idea of somehow maintaining all backward compatibility.
The result must have been a regression 😅
I will try to work on improvements based on the code you suggested.

t-hamano committed:
Fix regressions

t-hamano commented:
I have tested your suggestion, but could not solve the only problem regarding the border color.

When I change the text color, the border color should be the same color. But when I change the text color from the global style, it retains the backward compatible $gray-300.


https://user-images.githubusercontent.com/54422211/190894326-ba1e8e79-391a-419c-8b6b-edd71b890a1a.mp4

I have done a lot of testing and I think there are two possible choices:

- Remove backward compatibility of border colors
- Allow current behavior

aaronrobertshaw reviewed (approved):
> When I change the text color, the border color should be the same color. But when I change the text color from the global style, it retains the backward compatible $gray-300.

This is understandable and a tricky one to solve given when styles are generated for global styles we have nothing to be able to hook onto to override or omit that default border color.

To be clear for others coming to this PR in future:
- Setting text color for individual blocks **does** result in the default border color being omitted and thus picking up the `currentColor` value.
- It is only the global styles or theme.json application of a text color that **does not** result in this behaviour.

The matching of the border color to the text color is a _new_ feature. As such I don't think it should trump breaking existing uses of this block out in the wild. Especially, when there is a workaround i.e. setting it on the individual block, even if tedious.

There's also another option. You could opt into border supports for this block. It would provide a means to set the border color at the Calendar block level in the Global Styles or theme.json. 

I don't think that needs to block this PR but is definitely something you could do in a follow-up to this.

Given all of that, I believe this PR is in a reasonable place to land. Thank you for your patience and continued iteration towards this end. 

Nice work @t-hamano 🚢 

---

✅ Theme.json - Colors applied via theme.json blocks config appear correctly
✅ Global Styles - Global Styles correctly override the theme.json values
✅ Individual Block Styles - Color values set at the individual block level take precedence above others
✅ The caption receives any background color that's been set
✅ The table header receives background color selection
✅ Without a background the table header gets a default grey background
✅ Original hardcoded default values for text, header background, and border colors are maintained.

t-hamano commented:
> Setting text color for individual blocks does result in the default border color being omitted and thus picking up the currentColor value.
> It is only the global styles or theme.json application of a text color that does not result in this behaviour.

I see, let's go with this!
Thank you for helping me resolve the issue, even though it was a complex one 😊
